### DISCUSSION TITLE
How to achieve accessing complex relational model data?

### DISCUSSION THREAD

**Discussion by JMoodyFWD** [phalcon/cphalcon#16075] [Q&A] [answered]
![Screenshot 2022-08-26 055506](https://user-images.githubusercontent.com/8059176/186898496-b07343ae-8869-49bf-87d7-3c31ada77249.jpg)

In my screenshot is a table structure and relationships I've laid out within my database to make a virtual mailbox. This is necessary for my medical app so we can provide nearly the same functionality like GMail, but without data ever leaving the app to an external client.

### A brief scope of the tables and their purpose:

`mail_thread` - Acts as the container for the messages sent within it.
`mail_thread_participants` - Each user involved with a mail thread is logged here, both to a.) only display threads they are involved with and b.) for admin logging and reporting.
`mail_messages` - All messages sent by a user is logged here and assigned to its respective thread.
`mail_message_recipients` - This is a further scope of permission to only show users messages within a thread where they were assigned as a recipient.
`mail_message_attachments` - All file attachments uploaded to our CDN gets logged here with the appropriate info to show within the message it belongs to.

### The relationship is as follows:

mail_thread
-----(One to Many) mail_thread_participants (linked by `mail_thread.id` to `mail_thread_participants.id`)
-----(One to Many) mail_message (linked by `mail_thread.id` to `mail_thread_participants.id`)
----------(One to Many) mail_message_participants (linked by `mail_message.id` to `mail_message_participants.message_id`)
----------(One to Many) mail_message_attachments (linked by `mail_message.id` to `mail_message_participants.message_id`)

### When a user opens their mailbox, I want the following to occur:

The database is queried for any rows within mail_thread and only returns any results where mail_thread has a relational mail_thread_recipients record with their user ID.

For each mail_thread record found, all related mail_message records will be gathered where mail_message has a relational mail_message_recipients record with their user ID.

For each message found, all related mail_message_attachments records will be gathered.

### So with that scope out of the way, my questions are as follows:

1.) Is this type of structure acceptable or is there a more efficient way of storing/fetching these records with the intended functionality provided?
2.) With this level of complexity, is using the find() method possible, or would I need to use Phalcon's query builder to manually join tables within a query?
3.) If the query builder is necessary, how would I traverse through the relational records after getting the mail_threads records gathered?

I've never stored data to this level of relational complexity with Phalcon, so I'm a bit lost on how to execute this properly. Any help would be greatly appreciated!

**Comment by niden** [comment, marked as answer]
My 2 cents

The structure is very nicely done, normalized. More on improving this below.

To get the data as efficiently as possible you will need to load the minimal amount of data that you want to display per action i.e. user clicks Inbox -> need to show all emails and who it was sent to.

This way you can query only the mail thread and recipients and display that information. When a user clicks on a mail thread, you can further go to display each message and whether it has attachments or not. You get the point. You minimize queries based on the amount of information you need to display.  Some minor reorganization might help you there also, for instance storing the number of attachments per message in the message table. This way you don't need to do a `count()` for attachments when displaying say the Inbox.

I would go with the query builder directly, to avoid n+1, unless this is the message display.

Message display: You just query one message record. Using that, you can use `getRelated()` to retrieve recipients, attachments etc. relating to that message. 

Non message display (say inbox): Use the query builder to retrieve your data. If necessary, query first to get the IDs of the satellite tables that relate with your records, then use `WHERE id IN (....)` for your main query. This will significantly reduce your queries.

Finally, once this grows to the point that queries become slower, you might want to invest in a text search engine (Elasticsearch for instance) which will store all your data denormalized. The index of each row in Elasticsearch will be updated with events such as `afterSave()` in each model, ensuring that both are in sync. When you query though, you do a text query in ES not in your database which will speed things up significantly.

There was an app we had many moons ago, which displayed a list of records. We had 17 joins in that model. Type as you go search was 5 seconds per character. Using Elasticsearch it was reduced to 0.1

**Reply by JMoodyFWD** [reply]
Wow, thanks @niden ! That was way more valuable of an answer than just 2 cents! I know this reply will be long, but I want to clarify with you one final time if you have a few moments to make sure I'm on the right track. I really appreciate you taking the time to help!

First, I'm glad that my database structure works out. This is not too dissimilar from how I've set up the rest of the 90+ tables in the app, so it's definitely reassuring that I'm on the right track.

I have attached an updated structure based on your input as well as some realizations that I had. Here is my reasoning if you wouldn't mind taking a look and see if this looks good to you:

![Screenshot 2022-08-27 052410](https://user-images.githubusercontent.com/8059176/187028045-aa8b8a71-1d88-40f3-90fd-c8b22583d025.jpg)

### mail_thread

- Added `total_attachment_count` and `total_attachment_file_size` columns to get the cumulative total of attachments and the total file size. This allows me to a.) indicate on the main inbox view whether a thread has attachments or not and b.) allow from an administrative side to determine the data usage of threads for evaluation of our service cost and usage and pricing model.

### mail_message

- Added an `attachment_count` column based on your suggestion to make it easier in the view to show the count of the attachments as well as determine if the view needs to loop through the attachments to display them.

### mail_message_participants

- Added an `is_read` column so we can adequately determine and display to a user if they have any unread messages. I would add a javascript function that when the message is in the viewport for more than 3-5 seconds, we will silently call back to our API to mark that message as read for the user.

As for the views themselves, this is what I'm thinking:

### Inbox View
Run a single query that grabs the last 10 records in `mail_thread` based on their inclusion in it using their user ID against the `mail_thread_participants` table, along with the latest message that the user is included on using their user ID against the `mail_message_recipients` table. This query would be used on the pagination functionality via AJAX where it would use the page numbers to calculate the record offset and update the list of threads.

Based on this query needing to join three tables to the initial `mail_thread` table, am I able to accomplish that with the Query Builder, or would I need to just execute a PHQL query? I read the documents but all the join examples only seem to use a single join, so I'm not sure if multiple joins are possible.

### Thread View
I would need to essentially do two queries on this page from what I'm seeing: the first would be to ensure that the user is in fact included in one of the records of `mail_thread_participants` and if so, proceed to grab the 5 latest messages via Query Builder where they are included in each message's `mail_message_recipients`. This query would be used to reveal older messages in a thread of a rate of 5 messages per query, in order to keep the query and data smaller. Each message displayed can check the `attachment_count` and loop through attachments if the result is greater than 0. When the user is viewing the messages, run an interval in javascript to check if they are viewing a message in their viewport that has a data-attribute of `is_read="0"`, and then send a silent call to an API to mark that message as read as well as reassigning that data attribute value to "1', which would keep them up to date on inbox view for unread messages.

---

Thanks again for your help and also thanks for recommending ElasticSearch! That's already been added on our list of to-dos once we start saturating the database with a high volume of records.

**Reply by niden** [reply]
Comments below:

> I have attached an updated structure based on your input as well as some realizations that I had. Here is my reasoning if you wouldn't mind taking a look and see if this looks good to you:

This looks just fine!

> - Added `total_attachment_count` and `total_attachment_file_size` columns to get the cumulative total of attachments and the total file size. This allows me to a.) indicate on the main inbox view whether a thread has attachments or not and b.) allow from an administrative side to determine the data usage of threads for evaluation of our service cost and usage and pricing model.

For admin stuff, it is not uncommon to duplicate the information contained in your main tables, so as not to burden such tables with admin operations. For instance, you will need to know how much space the user uses so that you can bill them if they go over. That number does not necessarily need to live only in the `mail_thread` table. You can have a `users` table that links to `admin_statistics` (for example) which will contain metadata for the user that the admin requires. That metadata will be updated when relevant operations occur. For instance an email came in with 1MB attachment, as I enter it in the database, I also update the `admin_statistics` table. The same with any other CRUD operations. This way you query only the "meta" table for users and not the main tables. Later on of course you can dump that table all together, when using Elasticsearch because at that point all the data will be denormalized and ready for you to pick up.

> ### mail_message_participants
> - Added an `is_read` column so we can adequately determine and display to a user if they have any unread messages. I would add a javascript function that when the message is in the viewport for more than 3-5 seconds, we will silently call back to our API to mark that message as read for the user.

This is a tough one. You will have degradation here no matter what you do. Scanning the relationship to see who read what is going to be costly. I would go with the approach of 2 additional tables. One will contain `user_id` and `unread_count`. This will give you very quickly the amount of messages that the user has unread. This result of course can be cached and furthermore it could come from the `admin_statistics` table (see above) so you can skip one table. The second table would be a `thread_id`, `user_id` group. The table will hold only the threads/messages that are unread and by which user. Once the user reads it, you delete the record. Yes the table can end up being massive, that is the drawback but it can also be mitigated later on to Elasticsearch. By using this secondary `mail_threads_unread` (or something like that) table, you can easily filter it with a relationship or a `WHERE IN (...)` clause to display only the unread emails (bold perhaps) in the current view/page.

> ### Inbox View
> Run a single query that grabs the last 10 records in `mail_thread` based on their inclusion in it using their user ID against the `mail_thread_participants` table, along with the latest message that the user is included on using their user ID against the `mail_message_recipients` table. This query would be used on the pagination functionality via AJAX where it would use the page numbers to calculate the record offset and update the list of threads.

The query should run once, store the data in a cache (say Redis) and keep it there until a new message arrives OR a change happens i.e. user deletes a message, reads a message etc. from that particular view. This way you are not hitting the DB every 10 seconds.

> Based on this query needing to join three tables to the initial `mail_thread` table, am I able to accomplish that with the Query Builder, or would I need to just execute a PHQL query? I read the documents but all the join examples only seem to use a single join, so I'm not sure if multiple joins are possible.

Yes multiple joins are possible and a lot easier when you use aliases. I would use the builder because it is very verbose in your code.

> ### Thread View
> I would need to essentially do two queries on this page from what I'm seeing: the first would be to ensure that the user is in fact included in one of the records of `mail_thread_participants` and if so, proceed to grab the 5 latest messages via Query Builder where they are included in each message's `mail_message_recipients`. This query would be used to reveal older messages in a thread of a rate of 5 messages per query, in order to keep the query and data smaller. Each message displayed can check the `attachment_count` and loop through attachments if the result is greater than 0. When the user is viewing the messages, run an interval in javascript to check if they are viewing a message in their viewport that has a data-attribute of `is_read="0"`, and then send a silent call to an API to mark that message as read as well as reassigning that data attribute value to "1', which would keep them up to date on inbox view for unread messages.

Check my idea above for the additional table.

If you have not done so already, make sure that you utilize a cache mechanism as much as possible. This will be used at least for now until you get ES going later on.

**Reply by JMoodyFWD** [reply]
Thanks again @niden all solid advice! We're gonna incorporate the `admin_statistics` like you advised and store copies of the pertinent data, and I've got Redis so that will help with the caching. ES will be phased in when we start growing to the point of it being cost-effective and impactful on the load.

I'll start planning out those suggestions you made regarding the `is_read` debacle. At the end of the day, there's no real great solution to implement, but we definitely need to give users from the inbox view and message view some kind of indication what is new for them from their last visit. Splitting it off to a separate table though and getting it cached is definitely the most cost effective implementation, so I'll get that figured out and polished. Also, because of this discussion, we've ultimately decided that messages that aren't flagged by users to archive will be auto-purged after 90 days, so that should help keep the tables lean enough that any kind of extra scanning should have its impact minimal on load times when factoring in Redis.

Thanks again for being such a huge help. We've been running on a shoestring budget and I'm the sole developer for what is way too big of an app for one person, so being able to get some guidance, especially from someone as busy as you, is something I'm eternally grateful for.

**Reply by niden** [reply]
Happy to help. All of the above have been issues that I have encountered so it was easy to offer pointers.

One thing that bit me in the past was cache invalidation so be careful on that. When your app pings the server (say for unread messages) it should do the db hit, store in the cache and send the numbers back. The next ajax call would be from the cache.

On a separate process of receiving an email, when the totals are updated, that initial cache should be invalidated so that the next ajax call would hit the db. You could also just update the cache instead of clearing it to avoid that db hit every X seconds or so.

One other thing that comes to mind is work with your payloads i.e. what is coming back to the front end from say ajax calls. What you could do is check payload interop (https://github.com/payload-interop/payload-interop). This is for the Action Domain Responder pattern or even Domain Driven Design. You could use it as your response vehicle so that your views (at least on the ajax/js side) receive a consistent object to work with. Another implementation could be a simple `Payload` class:

```php
<?php

namespace App\Http;

use function is_array;

class Payload
{
    public const CODE_SUCCESS           = 2000;
    public const CODE_ERROR             = 3000;
    public const CODE_INCORRECT_PAYLOAD = 3004;
    public const CODE_NOT_FOUND         = 3404;

    public const STATUS_SUCCESS = 'success';
    public const STATUS_ERROR   = 'error';

    private $message = '';
    private $code    = 2000;
    private $status  = 'success';
    private $payload = [];

    /**
     * Gets the payload code
     *
     * @return int
     */
    public function getCode(): int
    {
        return $this->code;
    }

    /**
     * Gets the payload message
     *
     * @return string
     */
    public function getMessage(): string
    {
        return $this->message;
    }

    /**
     * Gets the payload
     *
     * @return array
     */
    public function getPayload(): array
    {
        return $this->payload;
    }

    /**
     * Gets the status
     *
     * @return string
     */
    public function getStatus(): string
    {
        return $this->status;
    }

    /**
     * Initializes the internal variables (for reusability)
     *
     * @return $this
     */
    public function init(): Payload
    {
        $this->message = '';
        $this->code    = 2000;
        $this->status  = 'success';
        $this->actions = [];
        $this->payload = [];

        return $this;
    }

    /**
     * Returns the formatted payload
     *
     * @return array<string,integer|string|array>
     */
    public function render()
    {
        return [
            'code'    => $this->getCode(),
            'status'  => $this->getStatus(),
            'message' => $this->getMessage(),
            'actions' => $this->getActions(),
            'payload' => $this->getPayload(),
        ];
    }

    /**
     * Sets the code
     *
     * @param int $code
     *
     * @return $this
     */
    public function setCode(int $code): Payload
    {
        $this->code = $code;

        return $this;
    }

    /**
     * Sets an error message, status and code
     *
     * @param string $message
     *
     * @return $this
     */
    public function setError(string $message): Payload
    {
        $this->code    = self::CODE_ERROR;
        $this->status  = self::STATUS_ERROR;
        $this->message = $message;

        return $this;
    }

    /**
     * Sets the message
     *
     * @param string $message
     *
     * @return $this
     */
    public function setMessage(string $message): Payload
    {
        $this->message = $message;

        return $this;
    }

    /**
     * Sets the status
     *
     * @param string $status
     *
     * @return $this
     */
    public function setStatus($status = self::STATUS_SUCCESS): Payload
    {
        $this->status = $status;

        return $this;
    }

    /**
     * Sets a success message, status and code
     *
     * @param string $message
     *
     * @return $this
     */
    public function setSuccess(string $message): Payload
    {
        $this->code    = self::CODE_SUCCESS;
        $this->status  = self::STATUS_SUCCESS;
        $this->message = $message;

        return $this;
    }

    /**
     * Sets the payload
     *
     * @param array|mixed $payload
     *
     * @return $this
     */
    public function setPayload($payload): Payload
    {
        $this->payload = (true === is_array($payload)) ?
            $payload :
            [(string) $payload];

        return $this;
    }
}
```

Returning:

```php
return $this->payload->render();
```

will give you the JSON array you need. (note you will have to set that in your base controller i.e. `if action = ajax -> setJsonContent blah blah`

Then your JS will have everything it needs i.e. the status, payload etc. You can also map those same status codes in your global JS code and once your ajax call returns, a function will decide what to do with the data (if error for instance). The above class can easily be extended to whatever you need to send back and forth.
